api/src/main/java/io/grpc/NameResolver.java

```diff
@@ -699,8 +723,20 @@ public static final class Builder {
        * Sets the addresses resolved by name resolution.  This field is required.
        *
        * @since 1.21.0
+       * @deprecated Will be superseded by setAddressesOrError
        */
+      @Deprecated
       public Builder setAddresses(List<EquivalentAddressGroup> addresses) {
+        this.addresses = StatusOr.fromValue(addresses);
+        return this;
+      }
+
+      /**
+       * Sets the addresses resolved by name resolution or the error in doing so. This field is
+       * required.
+       * @param addresses Resolved addresses or an error in resolving addresses
+       */
+      public Builder setAddressesOrError(StatusOr<List<EquivalentAddressGroup>> addresses) {
         this.addresses = addresses;
```

Comment:
Add `checkNotNull(addresses)` here.

Hint: for the deprecated method have it call the new method, even though it is trivial to inline, simply so that we don't have to think about whether it also needs the checks.